File: 381-insert-delete-getrandom-o1-duplicates-allowed/insert-delete-getrandom-o1-duplicates-allowed.py
import random
from collections import defaultdict

class RandomizedCollection:

    def __init__(self):
        self.nums = []
        self.idx = defaultdict(set)

    def insert(self, val: int) -> bool:
        not_present = len(self.idx[val]) == 0

        self.nums.append(val)
        self.idx[val].add(len(self.nums) - 1)

        return not_present

    def remove(self, val: int) -> bool:
        if not self.idx[val]:
            return False

        remove_idx = self.idx[val].pop()
        last_val = self.nums[-1]

        if remove_idx != len(self.nums) - 1:
            self.nums[remove_idx] = last_val

            self.idx[last_val].remove(len(self.nums) - 1)
            self.idx[last_val].add(remove_idx)

        self.nums.pop()

        return True

    def getRandom(self) -> int:
        return random.choice(self.nums)

File: 381-insert-delete-getrandom-o1-duplicates-allowed/insert-delete-getrandom-o1-duplicates-allowed.java
import java.util.*;

class RandomizedCollection {

    private ArrayList<Integer> list;
    private HashMap<Integer, Set<Integer>> map;
    private Random random;

    public RandomizedCollection() {
        list = new ArrayList<>();
        map = new HashMap<>();
        random = new Random();
    }

    public boolean insert(int val) {
        boolean notPresent = !map.containsKey(val);

        map.putIfAbsent(val, new HashSet<>());
        map.get(val).add(list.size());

        list.add(val);

        return notPresent;
    }

    public boolean remove(int val) {
        if (!map.containsKey(val))
            return false;

        // Get one occurrence of val
        int removeIndex = map.get(val).iterator().next();

        // Remove this index from val's set
        map.get(val).remove(removeIndex);

        int lastIndex = list.size() - 1;
        int lastValue = list.get(lastIndex);

        if (removeIndex != lastIndex) {
            // Move last value to removed position
            list.set(removeIndex, lastValue);

            // Update lastValue's index set
            map.get(lastValue).remove(lastIndex);
            map.get(lastValue).add(removeIndex);
        }

        list.remove(lastIndex);

        if (map.get(val).isEmpty()) {
            map.remove(val);
        }

        return true;
    }

    public int getRandom() {
        return list.get(random.nextInt(list.size()));
    }
}